Refactoring to suppress eslint warnings

kohsheen1234 opened the pull request:
Enabling the eslint/no-param-reassign throws some warnings with the existing code.

Made necessary changes to suppress the warnings

Fixes #3927

Signed-off-by: Kohsheen Tiku <kohsheen.t@gmail.com>

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/master/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/master/CONTRIBUTING.md#committing
-->

kohsheen1234 commented on web/app/js/components/MetricsTable.jsx:
```diff
@@ -217,7 +217,8 @@ const preprocessMetrics = metrics => {
     _each(datum.latency, (value, quantile) => {
-      datum[quantile] = value;
+      const datum_ = datum;
+      datum_[quantile] = value;
```
@cpretzer I am not very sure of this case, but I guess it would work as expected. :)

kohsheen1234 replied:
@Pothulapati and @cpretzer . There is no need to return 'datum' . I have verified this by interchanging datum by datum_ and the function works as expected.

` const preprocessMetrics = metrics => {
  const tableData = _cloneDeep(metrics);
  _each(tableData, datum_ => {
    _each(datum_.latency, (value, quantile) => {
      const datum = datum_;
      datum[quantile] = value;
    });
  });
  return tableData;
};
`
So, I do not see any need of making further changes. :)

alpeb replied:
After having added the `"no-param-reassign": ["error", { "props": false }]` rule, you no longer need to edit this file, nor `Tap.jsx`  :wink:

kohsheen1234 commented:
@Pothulapati I've done my rebase. Please have a look

kohsheen1234 commented:
@Pothulapati and @cpretzer . There is no need to return 'datum' . I have verified this by interchanging datum by datum_ and the function works as expected.

` 
const preprocessMetrics = metrics => {
  const tableData = _cloneDeep(metrics);
  _each(tableData, datum_ => {
    _each(datum_.latency, (value, quantile) => {
      const datum = datum_;
      datum[quantile] = value;
    });
  });
  return tableData;
};
`

So, I do not see any need of making further changes. Also I have done my rebase, there are no conflicts any more :)

cpretzer commented:
@kohsheen1234 thanks this the note! 

I observed the same behavior when I tested the code and was looking to understand why it works that way. I _think_ it has to do with how variables are assigned.

Anyway, everything works, so thanks for the PR! 🙌

alpeb commented on web/app/.eslintrc:
```diff
@@ -84,6 +84,6 @@
     // These rules represent best practices that we're not adhering to, and should be enabled
     "jsx-a11y/click-events-have-key-events": 0,
-    "no-param-reassign": 0
+    "no-param-reassign": 1
```
I think it's better to just delete this line to enable that rule

kohsheen1234 replied:
@alpeb  I will do the same, thank you for the suggestion :)

kohsheen1234 replied:
By using this line `"no-param-reassign": ["error", { "props": false }] ` , we can reassign the properties of the objects that are passed as parameters.

alpeb replied:
That looks good :+1:

alpeb commented on web/app/js/components/ErrorModal.jsx:
```diff
@@ -52,20 +52,21 @@ class ErrorModal extends React.Component {
 
     const byPodAndContainer = _map(errorsByPod, (podErrors, pod) => {
       const byContainer = _reduce(podErrors.errors, (errors, err) => {
+        const errors__ = errors;
```
nit: could you use a single underscore, like you did in your other changes?

kohsheen1234 replied:
Surely, I'll use a single underscore. :)

alpeb commented:
Actually, I take my previous comment back :blush: 
When doing `const datum_ = datum;` the new var receives a reference, not a copy of `datum`, so mutating it does mutate the original one.
That being said, I believe creating new variables is sort of missing the spirit of the rule. I think it's fine, when the argument is an object, to allow to modify its properties. So I would suggest we relax the rule with `{ "props": true }`. What do you think?

cpretzer commented:
> Actually, I take my previous comment back 😊
> When doing `const datum_ = datum;` the new var receives a reference, not a copy of `datum`, so mutating it does mutate the original one.
> That being said, I believe creating new variables is sort of missing the spirit of the rule. I think it's fine, when the argument is an object, to allow to modify its properties. So I would suggest we relax the rule with `{ "props": true }`. What do you think?

👍 I think that makes sense and makes for more readable/understandable code.

kohsheen1234 commented:
> Actually, I take my previous comment back 😊
> When doing `const datum_ = datum;` the new var receives a reference, not a copy of `datum`, so mutating it does mutate the original one.
> That being said, I believe creating new variables is sort of missing the spirit of the rule. I think it's fine, when the argument is an object, to allow to modify its properties. So I would suggest we relax the rule with `{ "props": true }`. What do you think?

By default, in the rule eslint/no-param-reassign we have, `{ "props": false}` , which means that we can directly assign values to the properties of the object which is passed as an argument. By making`{"prop":true}`, we will simply make the rule more strict instead of making it relaxed.

![params](https://user-images.githubusercontent.com/22408440/74053742-79ad1e00-4a02-11ea-89e6-ab0b5fb91e4e.PNG)

alpeb commented:
> By default, in the rule eslint/no-param-reassign we have, `{ "props": false}` , which means that we can directly assign values to the properties of the object which is passed as an argument. By making`{"prop":true}`, we will simply make the rule more strict instead of making it relaxed.

You're right. But then I'm puzzled: why was something like `NetworkGraph.jsx` giving warnings if we were just changing object properties?

kohsheen1234 commented:
> > By default, in the rule eslint/no-param-reassign we have, `{ "props": false}` , which means that we can directly assign values to the properties of the object which is passed as an argument. By making`{"prop":true}`, we will simply make the rule more strict instead of making it relaxed.
> 
> You're right. But then I'm puzzled: why was something like `NetworkGraph.jsx` giving warnings if we were just changing object properties?

@alpeb Even I got confused, I don't know why it shows warnings. But I realized that using this line `"no-param-reassign":` ["error", { "props": false }]` does not show those warnings. 

Also, I had a doubt: After making changes by assigning values to properties of the object, I feel the dashboard isn't working as expected, though I am not sure about the same.

This is what I get after making the necessary changes. The first one is showing 0/2.

![deploy](https://user-images.githubusercontent.com/22408440/74060439-fe526900-4a0f-11ea-9cee-6850a23f9fee.PNG)

alpeb commented:
Thanks fine :slightly_smiling_face: 
It just means the `coredns` pod has two replicas, none of which has been injected by linkerd, which is expected since none of the pods from the kube-system namespace should be injected.

kohsheen1234 commented:
@alpeb Thank you for the clarification. 👍  I have made changes as requested in the Pull Request. Can you just check if any further changes are required. :)

alpeb commented on web/app/js/components/TopModule.jsx:
```diff
@@ -202,37 +202,38 @@ class TopModule extends React.Component {
   }
 
   indexTopResult = (d, topResults) => {
+    const topResults_ = topResults;
```
This change is no longer required either

kohsheen1234 replied:
Yes. You are right. :) Just made the changes.

kohsheen1234 commented:
@alpeb Can you check now. I have made the changes you told. Thank you very much for helping me with all these changes :)

alpeb commented on web/node_modules/.yarn-integrity:
```diff
@@ -0,0 +1,10 @@
+  "lockfileEntries": {},
+  "files": [],
+  "artifacts": {}
+}
```
Could you please remove this file?

kohsheen1234 replied:
Done :)

alpeb commented on web/yarn.lock:
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
+
+
```
Could you also please remove this file?

kohsheen1234 replied:
Done

alpeb commented on web/app/js/components/util/Utils.js:
```diff
@@ -119,10 +121,11 @@ export const toClassName = name => {
+  let input_ = input;
   // make input lower case and strip out unwanted characters
-  input = input.replace(/[^A-Z0-9/.\-_*]/gi, '').toLowerCase();
+  input_ = input_.replace(/[^A-Z0-9/.\-_*]/gi, '').toLowerCase();
```
I think you can transform these into a single liner:
```javascript
let input_ = input.replace(/[^A-Z0-9/.\-_*]/gi, '').toLowerCase();
```

kohsheen1234 replied:
Done :)

kohsheen1234 commented:
> Thanks @kohsheen1234 !
> I've just got a few more minor comments, and then this should be ready to merge

@alpeb Thank you very much. :) . I made the new changes as well.

kohsheen1234 commented:
> Thanks @kohsheen1234 for addressing the feedback so quickly 
> 
> As a first time contributor we'd love to send you some swag! Do you mind filling out this [form](https://docs.google.com/forms/d/19GLBNfu8XnIETrwEI3ijoWlocAiDp3-j7CW3l4cTnyc/viewform)?

Thank you very much @alpeb for your guidance. I have filled the form for some swag. :+1:
